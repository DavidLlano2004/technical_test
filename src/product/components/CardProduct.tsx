// CardProduct.tsx
import { Icons } from "../../assets/icons/IconProvider";
import { formatPrice } from "../../helpers/truncatePrice";
import { Check } from "../../shared/components/check/Check";
import type { Product } from "../interfaces/product";

interface Props {
  product: Product;
  isSelected: boolean;
  onSelect: (productId: string) => void;
}

const productIconMap: Record<string, keyof typeof Icons> = {
  MFUND: "IconMFUND",
  CREA: "IconCREA",
  FICS: "IconFICS",
  BOLT: "IconBOLT",
};

const getProductColor = (productName: string): string => {
  const colorMap: Record<string, string> = {
    MFUND: "bg-green-dark-custom",
    CREA: "bg-blue-light-custom",
    FICS: "bg-blue-dark-custom",
    BOLT: "bg-orange-light-custom",
  };

  return colorMap[productName] || "bg-gray-500";
};

export const CardProduct = ({ product, isSelected, onSelect }: Props) => {
  const iconKey =
    product.iconProduct || productIconMap[product.nameProduct] || "IconDefault";
  const colorClass = getProductColor(product.nameProduct);

  const handleCardClick = () => {
    onSelect(product.numberProduct);
  };

  const handleCheckChange = (checked: boolean) => {
    if (checked) {
      onSelect(product.numberProduct);
    } else {
      onSelect(""); // Deseleccionar
    }
  };

  return (
    <div
      className={`w-full bg-white h-[400px] rounded-lg flex-col flex justify-between overflow-hidden transition ease-in duration-150 shadow cursor-pointer ${
        isSelected
          ? "border-[0.5px] border-primary-custom shadow-lg shadow-primary-custom/30"
          : "border-[0.5px] border-[#d8d8d8] hover:border-primary-custom/50 hover:shadow-primary-custom"
      }`}
      onClick={handleCardClick}
    >
      <div className="p-4 flex flex-col justify-between h-full">
        <div>
          <div className="flex justify-end items-center">
            <div onClick={(e) => e.stopPropagation()}>
              <Check isChecked={isSelected} onChange={handleCheckChange} />
            </div>
          </div>
          <div className="px-12">
            <p className="text-gray-custom-text font-bold text-lg">
              {product?.nameProduct}
            </p>
            <p className="text-gray-custom-text text-lg">
              #{product?.numberProduct}
            </p>
            <div className="h-[0.5px] bg-gray1-custom my-5"></div>
            <p className="text-gray-custom-text font-bold text-lg">
              Ya cuentas con:
            </p>
            <p className="text-lg text-primary-custom font-semibold">
              ${formatPrice(Number(product?.balanceProduct))}
            </p>
          </div>
        </div>
        <div>
          <img
            className="mx-auto w-10 translate-y-3"
            src={Icons[iconKey as keyof typeof Icons]}
            alt={product.nameProduct}
          />
          <div className="bg-gray2-custom p-8 rounded-md">
            <p className="text-center text-gray-custom-text font-semibold">
              {product.detaildProduct}
            </p>
          </div>
        </div>
      </div>
      <div className={`h-1.5 ${colorClass}`}></div>
    </div>
  );
};
