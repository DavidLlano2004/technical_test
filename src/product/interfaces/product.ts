export interface Product {
  nameProduct: string;
  numberProduct: string;
  balanceProduct: string;
  detaildProduct: string;
  iconProduct?: string;
}

export interface ProductsResponse {
  listCard: Product[];
}