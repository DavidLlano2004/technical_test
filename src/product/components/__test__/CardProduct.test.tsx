import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { CardProduct } from '../CardProduct';
import { Product } from '../../interfaces/product';

// Mock de los m칩dulos externos
jest.mock('../../../assets/icons/IconProvider', () => ({
  Icons: {
    IconMFUND: '/mock-icon-mfund.svg',
    IconCREA: '/mock-icon-crea.svg',
    IconFICS: '/mock-icon-fics.svg',
    IconBOLT: '/mock-icon-bolt.svg',
    IconDefault: '/mock-icon-default.svg'
  }
}));

jest.mock('../../../helpers/truncatePrice', () => ({
  formatPrice: jest.fn((price: number) => price.toLocaleString())
}));

jest.mock('../../../shared/components/check/Check', () => ({
  Check: ({ isChecked, onChange }: { isChecked: boolean; onChange: (checked: boolean) => void }) => (
    <input
      type="checkbox"
      data-testid="check-component"
      checked={isChecked}
      onChange={(e) => onChange(e.target.checked)}
    />
  )
}));

describe('CardProduct', () => {
  const mockProduct: Product = {
    numberProduct: 'PROD123',
    nameProduct: 'MFUND',
    balanceProduct: '50000',
    detaildProduct: 'Fondo de inversi칩n mutuo',
    iconProduct: undefined
  };

  const mockOnSelect = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Renderizado b치sico', () => {

    test('renders product icon with correct alt text', () => {
      render(
        <CardProduct
          product={mockProduct}
          isSelected={false}
          onSelect={mockOnSelect}
        />
      );

      const icon = screen.getByAltText('MFUND');
      expect(icon).toBeInTheDocument();
      expect(icon).toHaveAttribute('src', '/mock-icon-mfund.svg');
    });

    test('renders check component', () => {
      render(
        <CardProduct
          product={mockProduct}
          isSelected={false}
          onSelect={mockOnSelect}
        />
      );

      expect(screen.getByTestId('check-component')).toBeInTheDocument();
    });
  });

  describe('Estados de selecci칩n', () => {
    test('applies selected styles when isSelected is true', () => {
      const { container } = render(
        <CardProduct
          product={mockProduct}
          isSelected={true}
          onSelect={mockOnSelect}
        />
      );

      const card = container.firstChild as HTMLElement;
      expect(card).toHaveClass('border-primary-custom');
      expect(card).toHaveClass('shadow-lg');
      expect(card).toHaveClass('shadow-primary-custom/30');
    });

    test('applies default styles when isSelected is false', () => {
      const { container } = render(
        <CardProduct
          product={mockProduct}
          isSelected={false}
          onSelect={mockOnSelect}
        />
      );

      const card = container.firstChild as HTMLElement;
      expect(card).toHaveClass('border-[#d8d8d8]');
      expect(card).toHaveClass('hover:border-primary-custom/50');
    });

    test('check component reflects isSelected state', () => {
      const { rerender } = render(
        <CardProduct
          product={mockProduct}
          isSelected={false}
          onSelect={mockOnSelect}
        />
      );

      let checkbox = screen.getByTestId('check-component') as HTMLInputElement;
      expect(checkbox.checked).toBe(false);

      rerender(
        <CardProduct
          product={mockProduct}
          isSelected={true}
          onSelect={mockOnSelect}
        />
      );

      checkbox = screen.getByTestId('check-component') as HTMLInputElement;
      expect(checkbox.checked).toBe(true);
    });
  });

  describe('Interacciones', () => {
    test('calls onSelect with product number when card is clicked', async () => {
      const user = userEvent.setup();
      const { container } = render(
        <CardProduct
          product={mockProduct}
          isSelected={false}
          onSelect={mockOnSelect}
        />
      );

      const card = container.firstChild as HTMLElement;
      await user.click(card);

      expect(mockOnSelect).toHaveBeenCalledTimes(1);
      expect(mockOnSelect).toHaveBeenCalledWith('PROD123');
    });

    test('calls onSelect with product number when check is checked', async () => {
      const user = userEvent.setup();
      render(
        <CardProduct
          product={mockProduct}
          isSelected={false}
          onSelect={mockOnSelect}
        />
      );

      const checkbox = screen.getByTestId('check-component');
      await user.click(checkbox);

      expect(mockOnSelect).toHaveBeenCalledTimes(1);
      expect(mockOnSelect).toHaveBeenCalledWith('PROD123');
    });

    test('calls onSelect with empty string when check is unchecked', async () => {
      const user = userEvent.setup();
      render(
        <CardProduct
          product={mockProduct}
          isSelected={true}
          onSelect={mockOnSelect}
        />
      );

      const checkbox = screen.getByTestId('check-component');
      await user.click(checkbox);

      expect(mockOnSelect).toHaveBeenCalledTimes(1);
      expect(mockOnSelect).toHaveBeenCalledWith('');
    });

    test('prevents event propagation when clicking on check', () => {
      render(
        <CardProduct
          product={mockProduct}
          isSelected={false}
          onSelect={mockOnSelect}
        />
      );

      const checkbox = screen.getByTestId('check-component');
      const clickEvent = new MouseEvent('click', { bubbles: true });
      
      fireEvent.click(checkbox, clickEvent);
      
      // El onSelect debe ser llamado solo una vez (por el check, no por el card)
      expect(mockOnSelect).toHaveBeenCalledTimes(1);
    });
  });

  describe('Mapeo de iconos y colores', () => {
    test.each([
      ['MFUND', 'IconMFUND', 'bg-green-dark-custom'],
      ['CREA', 'IconCREA', 'bg-blue-light-custom'],
      ['FICS', 'IconFICS', 'bg-blue-dark-custom'],
      ['BOLT', 'IconBOLT', 'bg-orange-light-custom']
    ])('uses correct icon and color for %s product', (productName, expectedIcon, expectedColor) => {
      const testProduct: Product = {
        ...mockProduct,
        nameProduct: productName
      };

      const { container } = render(
        <CardProduct
          product={testProduct}
          isSelected={false}
          onSelect={mockOnSelect}
        />
      );

      const icon = screen.getByAltText(productName);
      expect(icon).toHaveAttribute('src', `/mock-icon-${expectedIcon.toLowerCase().replace('icon', '')}.svg`);

      const colorBar = container.querySelector(`.${expectedColor}`);
      expect(colorBar).toBeInTheDocument();
    });

    test('uses default color for unknown product', () => {
      const testProduct: Product = {
        ...mockProduct,
        nameProduct: 'UNKNOWN'
      };

      const { container } = render(
        <CardProduct
          product={testProduct}
          isSelected={false}
          onSelect={mockOnSelect}
        />
      );

      const colorBar = container.querySelector('.bg-gray-500');
      expect(colorBar).toBeInTheDocument();
    });
  });


  describe('Accesibilidad', () => {
    test('card has cursor pointer class', () => {
      const { container } = render(
        <CardProduct
          product={mockProduct}
          isSelected={false}
          onSelect={mockOnSelect}
        />
      );

      const card = container.firstChild as HTMLElement;
      expect(card).toHaveClass('cursor-pointer');
    });

    test('icon has proper alt attribute', () => {
      render(
        <CardProduct
          product={mockProduct}
          isSelected={false}
          onSelect={mockOnSelect}
        />
      );

      const icon = screen.getByAltText('MFUND');
      expect(icon).toBeInTheDocument();
    });
  });

  describe('Casos edge', () => {
    test('handles missing product properties gracefully', () => {
      const incompleteProduct = {
        numberProduct: 'PROD123',
        nameProduct: 'TEST',
        balanceProduct: '0',
        detaildProduct: 'Test product'
      } as Product;

      expect(() => {
        render(
          <CardProduct
            product={incompleteProduct}
            isSelected={false}
            onSelect={mockOnSelect}
          />
        );
      }).not.toThrow();
    });

    test('handles zero balance', () => {
      const zeroBalanceProduct: Product = {
        ...mockProduct,
        balanceProduct: '0'
      };

      render(
        <CardProduct
          product={zeroBalanceProduct}
          isSelected={false}
          onSelect={mockOnSelect}
        />
      );

      expect(screen.getByText('$0')).toBeInTheDocument();
    });
  });

  describe('Snapshot', () => {
    test('matches snapshot for unselected state', () => {
      const { container } = render(
        <CardProduct
          product={mockProduct}
          isSelected={false}
          onSelect={mockOnSelect}
        />
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    test('matches snapshot for selected state', () => {
      const { container } = render(
        <CardProduct
          product={mockProduct}
          isSelected={true}
          onSelect={mockOnSelect}
        />
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});