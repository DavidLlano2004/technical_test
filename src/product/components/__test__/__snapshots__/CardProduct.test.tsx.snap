// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CardProduct Snapshot matches snapshot for selected state 1`] = `
<div
  class="w-full bg-white h-[400px] rounded-lg flex-col flex justify-between overflow-hidden transition ease-in duration-150 shadow cursor-pointer border-[0.5px] border-primary-custom shadow-lg shadow-primary-custom/30"
>
  <div
    class="p-4 flex flex-col justify-between h-full"
  >
    <div>
      <div
        class="flex justify-end items-center"
      >
        <div>
          <input
            checked=""
            data-testid="check-component"
            type="checkbox"
          />
        </div>
      </div>
      <div
        class="px-12"
      >
        <p
          class="text-gray-custom-text font-bold text-lg"
        >
          MFUND
        </p>
        <p
          class="text-gray-custom-text text-lg"
        >
          #
          PROD123
        </p>
        <div
          class="h-[0.5px] bg-gray1-custom my-5"
        />
        <p
          class="text-gray-custom-text font-bold text-lg"
        >
          Ya cuentas con:
        </p>
        <p
          class="text-lg text-primary-custom font-semibold"
        >
          $
          50.000
        </p>
      </div>
    </div>
    <div>
      <img
        alt="MFUND"
        class="mx-auto w-10 translate-y-3"
        src="/mock-icon-mfund.svg"
      />
      <div
        class="bg-gray2-custom p-8 rounded-md"
      >
        <p
          class="text-center text-gray-custom-text font-semibold"
        >
          Fondo de inversión mutuo
        </p>
      </div>
    </div>
  </div>
  <div
    class="h-1.5 bg-green-dark-custom"
  />
</div>
`;

exports[`CardProduct Snapshot matches snapshot for unselected state 1`] = `
<div
  class="w-full bg-white h-[400px] rounded-lg flex-col flex justify-between overflow-hidden transition ease-in duration-150 shadow cursor-pointer border-[0.5px] border-[#d8d8d8] hover:border-primary-custom/50 hover:shadow-primary-custom"
>
  <div
    class="p-4 flex flex-col justify-between h-full"
  >
    <div>
      <div
        class="flex justify-end items-center"
      >
        <div>
          <input
            data-testid="check-component"
            type="checkbox"
          />
        </div>
      </div>
      <div
        class="px-12"
      >
        <p
          class="text-gray-custom-text font-bold text-lg"
        >
          MFUND
        </p>
        <p
          class="text-gray-custom-text text-lg"
        >
          #
          PROD123
        </p>
        <div
          class="h-[0.5px] bg-gray1-custom my-5"
        />
        <p
          class="text-gray-custom-text font-bold text-lg"
        >
          Ya cuentas con:
        </p>
        <p
          class="text-lg text-primary-custom font-semibold"
        >
          $
          50.000
        </p>
      </div>
    </div>
    <div>
      <img
        alt="MFUND"
        class="mx-auto w-10 translate-y-3"
        src="/mock-icon-mfund.svg"
      />
      <div
        class="bg-gray2-custom p-8 rounded-md"
      >
        <p
          class="text-center text-gray-custom-text font-semibold"
        >
          Fondo de inversión mutuo
        </p>
      </div>
    </div>
  </div>
  <div
    class="h-1.5 bg-green-dark-custom"
  />
</div>
`;
